### PR TITLE
[RFC] Dev container support

### PR DESCRIPTION
**Pull request by rynewang**
## Why are these changes needed?

Adds vscode dev container support. This provides an hermetic environment with all needed dependencies.

## How to use

1. If you are on mac: install "Docker for mac"
2. prepare a clean git environment.
3. open vscode, install "Dev Containers" extension.
4. open the workspace with dev containers
5. Now you are in a container with proper environments.

## Sharp edges

- If some C++ headers from deps are "not found": bazel build your pkg first.
- I used a `-aarch64` base image for my M1 mac.
- I can't get `bazel build //:ray_pkg` to work, it complains something like `python/ray/serve/generated/sedXhx0Ry` permission denied.

### CODE REVIEW

**Comment by scv119** [comment]
@aslonnie @can-anyscale any opinion on this PR?

also @rynewang, is there anyway we can test the correctness of the container in CI?

File: docker/devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# The devcontainer Dockerfile to set up an hermetic enviornment with all needed dependencies.
+
+FROM rayproject/ray-deps:2023-06-26-cu116-aarch64
```
Comment (aslonnie) [inline docker/devcontainer/Dockerfile:3]: how would you update this? this is not building from source.

for example HEAD's default gpu image is already cu118 now.

also, when this is on mac apple silicon, I do not know why cuda is being used. as I don't think cuda driver would work.

File: docker/devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+COPY . /ray
+# Install dependencies needed to build ray
+RUN sudo apt-get update && sudo apt-get install -y curl unzip cmake gcc g++ && sudo apt-get clean
+RUN sudo chown -R ray:users /ray && cd /ray && git init && ./ci/env/install-bazel.sh --system
```
Comment (aslonnie) [inline docker/devcontainer/Dockerfile:7]: local dev should not be using the `ci` scripts. I think what you need is just to install bazelisk.

**Review by aslonnie — COMMENTED** [review]
sorry, but I do not really understand what pain point this is trying to address. the container is not built from source, and it is developing in an arm64 linux container (which runs inside a linux vm), which is an environment that no one really uses today (and we also have almost no tests running on arm64 linux platforms).

note that, the requirements and dependencies are checked in inside the repo today, and gets updated like every week by people. so if the dev container here is not able to pick up the dependencies from the repo, I do not think it can really claim that "this provides a hermetic environment with all needed dependencies.".

also `ray-deps` do not have ml dependencies, it is just a `base-deps` with a wheel copied in. so the use seems to be a bit limited.

I feel that one can just create an arm64 graviton linux ec2 vm (or x86), and achieve the approximately same thing with vscode+ssh, and even without containers (which is basically the workflow I uses today).

maybe it makes linux local dev easier, but I worry about the false signals that it might provide. because the platform it uses is the least used one.

constructively, here are some points for improvement:

- if the purpose is to test with the dependencies, to make this useful, I think this should be building from source, ideally without adding new `Dockerfile` that is only used for dev.
- ray supposedly also supports osx and apple silicon, so if the purpose is to test ray library logic, instead of using a container, maybe we should get the apple silicon native dev environment work properly?
- for getting the dependencies, the requirements are also declared in `python/requirements.txt`. maybe just use that, rather than using `ray-deps`?

**Comment by aslonnie** [comment]
tl;dr, I do not fully understand why we need this. like if one gets a ubuntu ec2 instance, setup a conda env, and `pip install -e python/ray`, it can give basically the same thing? and it is:

- it will be testing the `x86_64` version, which is the most widely used one.
- and does not need a container?
